### DISCUSSION TITLE
Customize plan generation

### DISCUSSION THREAD

**Discussion by matteocarcano** [frePPLe/frepple#673] [Q&A]
Hi,
if i would customize the plan generation, adding new planning step after the tasks generation and the standard planning, what would be the correct sequence when the task should be executed? Because i need Frepple to pick up changes to start and end task's dates.

Thank you in advance

**Comment by jdetaeye** [comment]
Have a look at the commands.py files:
https://github.com/frePPLe/frepple/blob/master/freppledb/execute/commands.py
https://github.com/frePPLe/frepple/blob/master/freppledb/forecast/commands.py
https://github.com/frePPLe/frepple/blob/master/freppledb/input/commands/load.py
https://github.com/frePPLe/frepple/blob/master/freppledb/output/commands.py

These files are executed by the planning engine during the plan generation, and it's pretty easy to code additional planning logic in here.

For best maintainability you best put your customizations in a custom app, as documented here: https://frepple.com/docs/current/developer-guide/creating-an-extension-app.php

**Reply by matteocarcano** [reply]
Hi, 
i've analyzed files but my problem is that we have a cron-job that runs at night and changes the operationplan startdate and enddate and reosurce_id in the operationplanresource (importing data from our external system) directly on database.

This is not reflected in the application (e.g. resource report ,inventory status and delay on operationplan, delay on sales order), while if I change the start date/end date from the application (from manufactoring order section) I see that in the "runplan" log some operations were performed (exports of data), and I can see it reflected in resource report, delay ecc...

we have two options:
- we change data in every table in our cron job (I would rather not do it)
- we find the command to run/code to invoke so that the changes are correctly propagated

is there such a functionality?
are they steps that i find in file https://github.com/frePPLe/frepple/blob/master/freppledb/output/commands.py?

Thank you in advance

**Reply by jdetaeye** [reply]
With the commands.py files you can indeed integrate such pre- and post-processing steps in your planning workflow.    Muuuuch better than an independently running cronjob.

**Reply by matteocarcano** [reply]
sorry, I think I explained it wrong, what I need is to invoke only the "post-processing" of the planning process after the cronjob is run (unfortunately I cannot embed our integration into Frepple, so I cannot run it before the planning for example)
 
 what I would like to do is, after the cronjob has written the new data on the operationplan, run the same "tail" operations that `runplan` does, so that my manual changes are propagated to the rest of Frepple (resource report etc..)
 
 so my questions are:
 - is there a way to only trigger the last phases of the planner (to avoid overwriting what the cronjob did), either from python code or from command line?
 - what are the phases I should trigger?
 
 thanks!

**Comment by jdetaeye** [comment]
Ok, I see...
If you update startdate, quantity and assigned resource, you can let frepple recompute the end date.
If the status is "confirmed", we'll compute the end date only if it's still empty.

This recalculation is included in the "supply plan" option. 
Why can't you just regenerate the complete plan?

**Reply by matteocarcano** [reply]
Because I don't want to regenerate the plan automatically and create new production orders; I just want to update the information about the existing MOs

**Reply by jdetaeye** [reply]
You'll need to run all steps, except this one: https://github.com/frePPLe/frepple/blob/3a3624e8d1c5f39c6939bd8ed41ea9c75aa57d1d/freppledb/execute/commands.py#L127
You skip the step by returning -1 from the getWeight method.

Your mileage will vary with the Community Edition:
If the manufacturing orders are infeasible at your new dates, these infeasibilities will remain in your plan.
In the Enterprise Edition, we have extra solver capabilities to "repair" the plan and keep it feasible.

A typical use case:
Some manufacturing orders scheduled for yesterday didn't finish (due to downtime on a machine, lazy operator, rework, inefficiencies, ...) and will complete only today.
The plan repair functionality in the Enterprise Edition will reschedule/move out the manufacturing orders of today to a new date for adjust for this delay.
The Community Edition will not do do this, and just display a capacity overload on today's plan.
